File: Android/app/src/main/java/com/callandt/snipemobile/data/api/SnipeITOAuthService.kt
package com.callandt.snipemobile.data.api

import android.net.Uri
import android.util.Base64
import com.callandt.snipemobile.BuildConfig
import com.callandt.snipemobile.debug.AppLog
import com.callandt.snipemobile.ui.util.L10n
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.concurrent.TimeUnit

/** Browser OAuth login. */
object SnipeITOAuthService {
    /** Redirect for Snipe-IT's public mobile OAuth client. */
    const val REDIRECT_SCHEME = "com.grokability.snipeitmobile"
    const val REDIRECT_URI = "com.grokability.snipeitmobile://home"

    private val http = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val userAgent: String
        get() = "SnipeMobile/${BuildConfig.VERSION_NAME} (Android)"

    sealed class Discovery {
        data class OAuth(val clientId: String) : Discovery()
        data object Unavailable : Discovery()
        data object Unreachable : Discovery()
    }

    data class AuthSession(
        val clientId: String,
        val verifier: String,
        val state: String,
        val authorizeUrl: String,
    )

    data class LoginResult(
        val baseUrl: String,
        val token: String,
    )

    class ServiceException(val kind: Kind) : Exception() {
        enum class Kind {
            InvalidUrl,
            Cancelled,
            MissingCode,
            StateMismatch,
            TokenExchangeFailed,
            OauthUnavailable,
            ConnectionFailed,
        }

        override val message: String
            get() = when (kind) {
                Kind.InvalidUrl -> L10n.string("api_validate_invalid_url")
                Kind.Cancelled -> L10n.string("login_cancelled")
                Kind.MissingCode, Kind.StateMismatch, Kind.TokenExchangeFailed ->
                    L10n.string("login_failed")
                Kind.OauthUnavailable -> L10n.string("login_oauth_unavailable")
                Kind.ConnectionFailed -> L10n.string("login_connection_error")
            }
    }

    suspend fun discover(rawBaseUrl: String): Discovery = withContext(Dispatchers.IO) {
        val base = SnipeApiClient.normalizeBaseUrl(rawBaseUrl)
        val url = runCatching { java.net.URI("$base/api/v1/client").toURL() }.getOrNull()
            ?: return@withContext Discovery.Unreachable
        if (url.protocol != "http" && url.protocol != "https") return@withContext Discovery.Unreachable
        if (url.host.isNullOrBlank()) return@withContext Discovery.Unreachable

        val request = Request.Builder()
            .url(url)
            .header("Accept", "application/json")
            .header("User-Agent", userAgent)
            .get()
            .build()
        try {
            http.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    AppLog.network("OAuth discover HTTP ${response.code}")
                    return@withContext Discovery.Unavailable
                }
                val body = response.body?.string().orEmpty()
                val json = runCatching { JSONObject(body) }.getOrNull()
                    ?: return@withContext Discovery.Unavailable
                val rawId = json.opt("client_id")
                val clientId = rawId?.toString()?.takeIf { it.isNotBlank() && it != "null" }
                if (clientId.isNullOrBlank()) return@withContext Discovery.Unavailable
                AppLog.network("OAuth client discovered")
                Discovery.OAuth(clientId)
            }
        } catch (e: Exception) {
            AppLog.network("OAuth discover failed ${e.javaClass.simpleName}")
            Discovery.Unreachable
        }
    }

    fun startSession(rawBaseUrl: String, clientId: String): AuthSession {
        val base = SnipeApiClient.normalizeBaseUrl(rawBaseUrl)
        val verifier = makeCodeVerifier()
        val state = makeCodeVerifier()
        val challenge = makeCodeChallenge(verifier)
        val uri = Uri.parse("$base/oauth/authorize").buildUpon()
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("client_id", clientId)
            .appendQueryParameter("redirect_uri", REDIRECT_URI)
            .appendQueryParameter("code_challenge", challenge)
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("state", state)
            .appendQueryParameter("prompt", "login")
            .build()
        return AuthSession(
            clientId = clientId,
            verifier = verifier,
            state = state,
            authorizeUrl = uri.toString(),
        )
    }

    fun authorizationCode(callback: Uri, expectedState: String): String {
        val returnedState = callback.getQueryParameter("state")
        if (!returnedState.isNullOrBlank() && returnedState != expectedState) {
            throw ServiceException(ServiceException.Kind.StateMismatch)
        }
        val error = callback.getQueryParameter("error")
        if (!error.isNullOrBlank()) {
            AppLog.network("OAuth authorize error=$error")
            throw ServiceException(ServiceException.Kind.TokenExchangeFailed)
        }
        val code = callback.getQueryParameter("code")
        if (code.isNullOrBlank()) {
            throw ServiceException(ServiceException.Kind.MissingCode)
        }
        return code
    }

    suspend fun completeSignIn(
        rawBaseUrl: String,
        session: AuthSession,
        callback: Uri,
    ): LoginResult = withContext(Dispatchers.IO) {
        val base = SnipeApiClient.normalizeBaseUrl(rawBaseUrl)
        val code = authorizationCode(callback, session.state)
        val accessToken = exchangeCode(base, session, code)
        AppLog.network("OAuth login complete")
        LoginResult(baseUrl = base, token = accessToken)
    }

    private fun exchangeCode(base: String, session: AuthSession, code: String): String {
        val body = FormBody.Builder()
            .add("grant_type", "authorization_code")
            .add("client_id", session.clientId)
            .add("code", code)
            .add("code_verifier", session.verifier)
            .add("redirect_uri", REDIRECT_URI)
            .build()
        val request = Request.Builder()
            .url("$base/oauth/token")
            .header("Accept", "application/json")
            .header("User-Agent", userAgent)
            .post(body)
            .build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                AppLog.network("OAuth token HTTP ${response.code}")
                throw ServiceException(ServiceException.Kind.TokenExchangeFailed)
            }
            val json = JSONObject(response.body?.string().orEmpty())
            val token = json.optString("access_token")
            if (token.isBlank()) throw ServiceException(ServiceException.Kind.TokenExchangeFailed)
            return token
        }
    }

    private fun makeCodeVerifier(): String {
        val bytes = ByteArray(32)
        SecureRandom().nextBytes(bytes)
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }

    private fun makeCodeChallenge(verifier: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(verifier.toByteArray(Charsets.US_ASCII))
        return Base64.encodeToString(digest, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }
}

File: IOS:iPadOS/SnipeMobile/API/SnipeITOAuthService.swift
import CryptoKit
import Foundation
import UIKit

/// Browser OAuth login.
@MainActor
final class SnipeITOAuthService {
    static let shared = SnipeITOAuthService()

    /// Redirect for Snipe-IT's public mobile OAuth client.
    static let redirectScheme = "com.grokability.snipeitmobile"
    static let redirectURI = "com.grokability.snipeitmobile://home"

    enum Discovery {
        case oauth(clientId: String)
        case unavailable
        case unreachable
    }

    struct LoginResult {
        let baseURL: String
        let token: String
    }

    enum ServiceError: LocalizedError, Equatable {
        case invalidURL
        case cancelled
        case missingCode
        case stateMismatch
        case tokenExchangeFailed
        case oauthUnavailable
        case connectionFailed

        var errorDescription: String? {
            switch self {
            case .invalidURL:
                return L10n.string("api_validate_invalid_url")
            case .cancelled:
                return L10n.string("login_cancelled")
            case .missingCode, .stateMismatch, .tokenExchangeFailed:
                return L10n.string("login_failed")
            case .oauthUnavailable:
                return L10n.string("login_oauth_unavailable")
            case .connectionFailed:
                return L10n.string("login_connection_error")
            }
        }
    }

    static func normalizeBaseURL(_ value: String) -> String {
        var trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }
        guard !trimmed.isEmpty else { return trimmed }

        let lower = trimmed.lowercased()
        if !lower.hasPrefix("http://"), !lower.hasPrefix("https://") {
            trimmed = "https://\(trimmed)"
        }

        let pathSuffixes = [
            "/index.php/account/api",
            "/account/api",
            "/index.php/api/v1",
            "/index.php/api",
            "/api/v1",
            "/api",
            "/index.php",
        ]
        for suffix in pathSuffixes {
            if trimmed.lowercased().hasSuffix(suffix) {
                trimmed = String(trimmed.dropLast(suffix.count))
                break
            }
        }
        while trimmed.hasSuffix("/") {
            trimmed.removeLast()
        }
        return trimmed
    }

    func discover(baseURL raw: String) async -> Discovery {
        let base = Self.normalizeBaseURL(raw)
        guard let url = URL(string: "\(base)/api/v1/client"),
              url.scheme == "http" || url.scheme == "https",
              url.host != nil
        else {
            return .unreachable
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 8
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(SnipeITAPIClient.userAgent, forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .unreachable
            }
            guard (200...299).contains(http.statusCode) else {
                AppLog.network("OAuth discover HTTP \(http.statusCode)")
                return .unavailable
            }
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return .unavailable
            }
            let clientId: String?
            if let value = json["client_id"] as? String {
                clientId = value
            } else if let value = json["client_id"] as? Int {
                clientId = String(value)
            } else if let value = json["client_id"] as? NSNumber {
                clientId = value.stringValue
            } else {
                clientId = nil
            }
            guard let clientId, !clientId.isEmpty else { return .unavailable }
            AppLog.network("OAuth client discovered")
            return .oauth(clientId: clientId)
        } catch {
            AppLog.network("OAuth discover failed kind=\(SnipeITAPIClient.connectionFailureKind(from: error).rawValue)")
            return .unreachable
        }
    }

    func signIn(baseURL raw: String, clientId: String) async throws -> LoginResult {
        let base = Self.normalizeBaseURL(raw)
        guard let authorizeBase = URL(string: "\(base)/oauth/authorize") else {
            throw ServiceError.invalidURL
        }

        let verifier = Self.makeCodeVerifier()
        let challenge = Self.makeCodeChallenge(verifier)
        let state = Self.makeCodeVerifier()

        var components = URLComponents(url: authorizeBase, resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "redirect_uri", value: Self.redirectURI),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state),
        ]
        guard let authURL = components?.url else { throw ServiceError.invalidURL }

        AppLog.network("OAuth browser login started")
        let callbackURL = try await startAuthSession(url: authURL)
        let code = try Self.authorizationCode(from: callbackURL, expectedState: state)
        let accessToken = try await exchangeCode(
            baseURL: base,
            clientId: clientId,
            code: code,
            verifier: verifier
        )
        AppLog.network("OAuth login complete")
        return LoginResult(baseURL: base, token: accessToken)
    }

    // MARK: - Private

    private func presentationWindow() -> UIWindow? {
        let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        return scenes.flatMap(\.windows).first(where: \.isKeyWindow)
            ?? scenes.flatMap(\.windows).first
    }

    private func startAuthSession(url: URL) async throws -> URL {
        do {
            return try await SnipeITOAuthBrowserSession.start(
                url: url,
                scheme: Self.redirectScheme,
                window: presentationWindow()
            )
        } catch is CancellationError {
            throw ServiceError.cancelled
        } catch {
            throw ServiceError.tokenExchangeFailed
        }
    }

    private static func authorizationCode(from callbackURL: URL, expectedState: String) throws -> String {
        var values: [String: String] = [:]
        let items = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)?.queryItems
            ?? URLComponents(string: callbackURL.absoluteString)?.queryItems
            ?? []
        for item in items {
            if let value = item.value, values[item.name] == nil {
                values[item.name] = value
            }
        }
        if values["code"] == nil, let query = callbackURL.query {
            for pair in query.split(separator: "&") {
                let parts = pair.split(separator: "=", maxSplits: 1).map(String.init)
                if parts.count == 2, values[parts[0]] == nil {
                    values[parts[0]] = parts[1].removingPercentEncoding ?? parts[1]
                }
            }
        }
        if let returnedState = values["state"], returnedState != expectedState {
            throw ServiceError.stateMismatch
        }
        if let error = values["error"] {
            AppLog.network("OAuth authorize error=\(error)")
            throw ServiceError.tokenExchangeFailed
        }
        guard let code = values["code"], !code.isEmpty else {
            throw ServiceError.missingCode
        }
        return code
    }

    private func exchangeCode(
        baseURL: String,
        clientId: String,
        code: String,
        verifier: String
    ) async throws -> String {
        guard let url = URL(string: "\(baseURL)/oauth/token") else {
            throw ServiceError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(SnipeITAPIClient.userAgent, forHTTPHeaderField: "User-Agent")

        var body = URLComponents()
        body.queryItems = [
            URLQueryItem(name: "grant_type", value: "authorization_code"),
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "code", value: code),
            URLQueryItem(name: "code_verifier", value: verifier),
            URLQueryItem(name: "redirect_uri", value: Self.redirectURI),
        ]
        request.httpBody = body.percentEncodedQuery?.data(using: .utf8)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw ServiceError.connectionFailed
        }

        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            AppLog.network("OAuth token HTTP \((response as? HTTPURLResponse)?.statusCode ?? -1)")
            throw ServiceError.tokenExchangeFailed
        }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["access_token"] as? String,
              !token.isEmpty
        else {
            throw ServiceError.tokenExchangeFailed
        }
        return token
    }

    private static func makeCodeVerifier() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).base64URLEncodedString()
    }

    private static func makeCodeChallenge(_ verifier: String) -> String {
        let hash = SHA256.hash(data: Data(verifier.utf8))
        return Data(hash).base64URLEncodedString()
    }
}

private extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
